### PR TITLE
Feature/issue 44 transport npe

### PR DESCRIPTION
**Pull request by yepher**
Fix for Issue #44

### CODE REVIEW

**Comment by kalnik-a-a** [comment]
Maybe it will be useful to throw Exception on response with code  != 204 (400, 404 etc.) as it is done with not empty responses?

**Comment by yepher** [comment]
I like the idea. In reality any 2xx response in SparkPost is good. 4xx is a problem in the request and 5xx is a backend issue.

**Comment by yepher** [comment]
`com.sparkpost.transport.RestConnection.receiveResponse(HttpURLConnection, Response)` is getting a little out of hand. I am going to refactor it to make it more readable.

**Comment by yepher** [comment]
@kalnik-a-a Can you take a look at the updated changes. Error handling was in a couple of places. I tried to central like the error flow. Also Added server messages (if it exists) to the exceptions as it might contain additional information that might help the user of the library.

File: libs/sparkpost-lib/src/main/java/com/sparkpost/transport/RestConnection.java [hunk 1]
```diff
@@ -302,12 +308,77 @@ private Response receiveResponse(HttpURLConnection conn, Response response) thro
...
     }
 
-    private Response receiveEmptyResponse(HttpURLConnection conn, Response response) throws SparkPostException {
+    // This is used to handle 2xx HTTP responses
+    private Response receiveErrorResponse(HttpURLConnection conn, Response response) throws SparkPostException, IOException {
         response.setRequestId(conn.getHeaderField("X-SparkPost-Request-Id"));
-        return response;
+
+        if (conn.getContentLength() == 0) {
+            throw new SparkPostIllegalServerResponseException(
```
Comment (kalnik-a-a) [inline libs/sparkpost-lib/src/main/java/com/sparkpost/transport/RestConnection.java]: SparkPostErrorServerResponseException seems to be more appropriate here. Empty response != Illegal server response.

**Comment by kalnik-a-a** [comment]
What about separate method for reading Response from InputStream? 
Now it is duplicated in 3 places(2 in receiveSuccessResponse and 1 in receiveErrorResponse).

Something like this:
```
private Response readResponseFromInputStream(InputStream inputStream, StringBuilder sb, Response response) throws IOException {

        BufferedReader rd = new BufferedReader(new InputStreamReader(inputStream, DEFAULT_CHARSET));
        String line = "";
        while ((line = rd.readLine()) != null) {
            sb.append(line);
        }
        response.setResponseBody(sb.toString());
        logger.error("Server Response:\n" + sb.toString() + "\n");

        return response;
    }
```

File: libs/sparkpost-lib/src/main/java/com/sparkpost/transport/RestConnection.java (hunk 1 above)
Comment (yepher) [inline libs/sparkpost-lib/src/main/java/com/sparkpost/transport/RestConnection.java]: Fixed

**Comment by yepher** [comment]
@kalnik-a-a Good point about duplication. I have been looking about better methods for this but not quite ready to put into this release. I will create a separate issue for this.

**Pull request merged**